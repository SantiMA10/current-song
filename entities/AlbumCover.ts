export interface AlbumCover {
  height: number
  url: string
  width: number
}
