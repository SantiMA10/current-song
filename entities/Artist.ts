export interface Artist {
  name: string;
}
