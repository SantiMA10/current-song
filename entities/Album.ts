import { AlbumCover } from './AlbumCover';

export interface Album {
  name: string;
  images: AlbumCover[];
}
