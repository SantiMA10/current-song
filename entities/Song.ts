import { Album } from './Album'
import { Artist } from './Artist'

export interface Song {
  album: Album
  artists: Artist[]
  name: string
  duration_ms: number // eslint-disable-line camelcase
  progress_ms: number // eslint-disable-line camelcase
}
