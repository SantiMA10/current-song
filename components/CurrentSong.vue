<template>
  <div class="marquee">
    <div class="w-full flex items-center justify-center">
      <img v-if="albumCover" class="h-screen z-10" :src="albumCover.url" />
      <h1 class="z-0 whitespace-no-wrap text-white">
        {{ song.name }} - {{ artist }}
      </h1>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

import { AlbumCover } from '../entities/AlbumCover'
import { Artist } from '../entities/Artist'
import { Song } from '../entities/Song'

export default Vue.extend({
  props: {
    song: {
      required: true,
      type: Object as () => Song
    }
  },
  computed: {
    albumCover(): AlbumCover | undefined {
      if (!this.song) {
        return undefined
      }

      const images: AlbumCover[] = this.song.album.images

      return images.find(image => image.width > 64 && image.width < 400)
    },
    artist() {
      if (!this.song) {
        return undefined
      }

      const artists: Artist[] = this.song.artists

      return artists.map(({ name }) => name).join(',')
    }
  }
})
</script>

<style>
.marquee {
  width: 100%;
  overflow: hidden;
}
.marquee h1 {
  font-size: 3em;
  width: 100%;
  height: 100%;
  line-height: 50px;
  transform: translateY(-50%);
  animation: marquee 15s linear infinite;
}
@keyframes marquee {
  0% {
    transform: translateX(100%);
  }
  100% {
    transform: translateX(-100%);
  }
}
</style>
