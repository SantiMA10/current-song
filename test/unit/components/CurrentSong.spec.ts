import { mount } from '@vue/test-utils';

import CurrentSong from '~/components/CurrentSong.vue';
import { SongBuilder } from '~/test/builder/SongBuilder';
import { AlbumCover } from '~/entities/AlbumCover';
import { Artist } from '~/entities/Artist';

describe('CurrentSong', () => {
  it('not renders the album cover if there is images', () => {
    const subject = mount(CurrentSong, {
      propsData: {
        song: new SongBuilder().getInstance()
      }
    });

    expect(subject.find('img').exists()).toBeFalsy();
  });

  it('renders the album cover with width between 64 and 400 pixels', () => {
    const images: AlbumCover[] = [
      { url: 'https://picsum.photos/10', width: 10, height: 10 },
      { url: 'https://picsum.photos/200', width: 200, height: 200 }
    ];

    const subject = mount(CurrentSong, {
      propsData: {
        song: new SongBuilder().with({ album: { images } }).getInstance()
      }
    });

    const image = subject.find('img');

    expect(image.exists()).toBeTruthy();
    expect(image.attributes('src')).toEqual('https://picsum.photos/200');
  });

  it('renders multiple artist names', () => {
    const artists: Artist[] = [
      {
        name: 'artist 1'
      },
      {
        name: 'artist 2'
      }
    ];

    const subject = mount(CurrentSong, {
      propsData: {
        song: new SongBuilder().with({ artists }).getInstance()
      }
    });

    const songComponent = subject.find('.qa-song');

    expect(songComponent.exists()).toBeTruthy();
    expect(songComponent.text()).toContain(
      artists.map(({ name }) => name).join(', ')
    );
  });

  it('renders one artist name', () => {
    const artists: Artist[] = [
      {
        name: 'artist 1'
      }
    ];

    const subject = mount(CurrentSong, {
      propsData: {
        song: new SongBuilder().with({ artists }).getInstance()
      }
    });

    const songComponent = subject.find('.qa-song');

    expect(songComponent.exists()).toBeTruthy();
    expect(songComponent.text()).toContain(artists[0].name);
  });

  it('renders the song name', () => {
    const song = new SongBuilder().getInstance();

    const subject = mount(CurrentSong, {
      propsData: {
        song
      }
    });

    const songComponent = subject.find('.qa-song');

    expect(songComponent.exists()).toBeTruthy();
    expect(songComponent.text()).toContain(song.name);
  });

  it('renders', () => {
    const subject = mount(CurrentSong, {
      propsData: {
        song: new SongBuilder().getInstance()
      }
    });

    expect(subject.element).toMatchSnapshot();
  });
});
