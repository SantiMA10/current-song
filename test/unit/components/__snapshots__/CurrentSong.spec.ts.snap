// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentSong renders 1`] = `
<div
  class="w-full flex items-center justify-center"
>
  <!---->
   
  <div
    class="marquee"
  >
    <h1
      class="z-0 whitespace-no-wrap text-white qa-song"
    >
      
      I'm a song - Artist
    
    </h1>
  </div>
</div>
`;
